README.md
# Milestone-2-Project-The-Movie-Gallery
deployed at https://movie-gallery-gl39.onrender.com/

The Movie Gallery project offers the opportunity for anyone to add, edit and delete movies in the collection.
This app was designed with the user in mind. We made a point to focus on making the app visually pleasing 
with a smooth transition from one screen to the next. Our ultimate goal is to maintain optimal backend functionality that translates to a seamless front-end. 

The project is designed using the MERN (MongoDB, Express, React, and Node) full-stack application developmental practices that implement full
CRUD (Create, Read, Update, and Delete) functionality.

## References:

[W3W Schools](https://www.w3schools.com)

[stack overflow](https://stackoverflow.com/)

[Rest-rant project from backend](https://github.com/michaelangelesz/project-REST-rant)

[GPT engine 3.5](chat.openai.com)
used to aid with formatting, CSS, CLI, and installation issues

## Extra resources:

[bootstrap](https://www.bootstrapcdn.com/)

## App background image source links:

[Home page cinema background](https://www.flickr.com/photos/sharkypics/50243040877/)

[Add a Movie view background](https://www.freepik.com/free-photo/movie-background-collage_33752481.htm#query=film%20background&position=8&from_view=keyword&track=ais)

[Gallery galaxy background](https://www.pexels.com/photo/milky-way-photography-2312040/)

[Show page background ‘red curtain’ source](https://www.npr.org/2023/03/09/1160731547/spelling-bee-high-school-theater)

All movie posters sourced from [movieposterdb.com](https://www.movieposterdb.com/) with the exception of:

[Gadar 2](https://www.imdb.com/title/tt15441054/)

[Kung Fu Panda](https://www.amazon.com/Kung-Panda-Widescreen-Jack-Black/dp/B001ECQ75A)

[The Super Mario Bros](https://www.thesupermariobros.movie/character-gallery/)  

[The Goonies](https://xl.moviepxl_59113_cad39c9e.jpgosterdb.com/05_09/1985/0089218/xl_51014_0089218_1e6b1771.jpg?v=2023-09-16%2011:48:49)

[Background for New.jsx page](https://img.freepik.com/free-photo/movie-background-collage_23-2149876028.jpg?size=626&ext=jpg&ga=GA1.1.1929040866.1695618859&semt=ais)

[Background for Edit.jsx page](https://img.freepik.com/premium-photo/film-projector-film-director-s-chair-dark-place-3d-rendering_651547-474.jpg?size=626&ext=jpg&ga=GA1.1.1929040866.1695618859&semt=ais)

## Future Features:

Rating system

Comment section

Sorting by year, genre and title


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

9965df5 (HEAD -> main, origin/main, origin/HEAD, sass-practive) Merge pull request #28 from stexie99/devBranch
b71cc73 (origin/devBranch, devBranch) added consistancy to view pages
559be03 Merge pull request #27 from stexie99/devBranch
7fe2c13 checking out
a77aba6 Update README.md
aa12fd5 Update README.md
c9f50f6 touched up the README
f1cd61f adjust edit year defaultValue prop
8c64732 updated README and home.jsx
7706e28 touchup error404 view
ee23dcc edit image
4e1f2ca Merge branch 'moreUpdates' into devBranch
a3c2027 (origin/moreUpdates) use google fonts Lato
c06d8fe (origin/testing, testing) fixed minor syntax errors
2680bea Merge pull request #26 from stexie99/pseditBranch
8a12056 (origin/pseditBranch) updated edit and new views background pic, no horiz scroll, added movies to seed
fc237b6 fixed syntax
630f8b0 Merge pull request #25 from stexie99/feature2
3e437f4 (origin/feature2) Merge branch 'devBranch' into feature2
7518e35 (feature2) fixed minor errors
29419d4 tweeked minor code
f82bc95 add subtle animation to image links in the gallery
4b959ac fix error404 image link
88dfe95 Merge pull request #24 from stexie99/changesAfterdev2
3a24f96 (origin/changesAfterdev2) Merge branch 'devBranch' into changesAfterdev2
ef6cac0 work with mandeep error404 merge
07f7cc5 Merge pull request #23 from stexie99/changesAfterdev
a8cdf16 (origin/changesAfterdev) Merge branch 'devBranch' into changesAfterdev
ccc080c user experience tweaks
5e80e8b Merge pull request #22 from stexie99/errorBranch
bc2c87e Merge pull request #21 from stexie99/revert-20-seedBranch
ff97290 (origin/revert-20-seedBranch) Revert "add more movie data"
77e37de (origin/errorBranch) add seed data,add 404,and css
8985311 undo small detail
86e9add Merge pull request #20 from stexie99/seedBranch
f4a31d6 add more movie data
90d1de1 update user experience
4d80df3 data/seed work
bcf0ef8 style adjustments, cleanup code
f3d8264 updated all controller to async await for better functionality and readability
56dffa7 add image
49c4779 tweeked minor features
fdde220 updated drop down menu
50d2a54 style tweaks
99d8005 touchup styles
efe473d added drop down menus for years
0cacd63 error image
88635a9 index view quick update
762ab60 Merge pull request #19 from stexie99/devBranch
2b8cf87 Merge pull request #18 from stexie99/commentBranch
666443c (origin/commentBranch) added comments to code
4c36036 Merge pull request #16 from stexie99/devBranch
e16617f updated schema to provide poster
737d6a4 update edit show route
b269469 changed controllers to using id, and added edit page
e873e1a adjusted css and added poster to new movies view
dbdd463 redirect routes
d13e65e Merge pull request #15 from stexie99/controllerBranch2
8e2c367 (origin/controllerBranch2, controllerBranch2) updated post route
7bab415 successfully updated delete route
f4d37fd touched up on routes and controllers
66a93fb Merge pull request #14 from stexie99/userExbranch2
2e7634f update seed file
38f5374 update CSS
105557c touched up branches
c548a9a fixed some controller routes
dc78c0e Merge pull request #13 from stexie99/controllerBranch
f3aa011 (origin/controllerBranch) Merge branch 'devBranch' into controllerBranch
ce6170c fixed new route
afbfd90 Merge pull request #12 from stexie99/userExbranch
ac19d16 (origin/userExbranch) Merge branch 'devBranch' into userExbranch
6491998 Merge pull request #11 from stexie99/newviews
76a238e Update style.css
835a036 update views & add navbar
5fe4444 update seed data
018a02d add show route
23e1267 make the movie posters appear
9038487 (origin/newviews) new movie working
8c096e3 changed index route to previous version
f54b2fe finished building basic structure of controllers
f50cca7 (origin/pstarEditBranch) Merge pull request #10 from stexie99/defaultView
3328920 add favicon
fc2f201 add some styling
879616a add movies index view
5c7224f add 404 page view
c740021 adding Home page view
ffd9d07  adding default view for layout
cb21fc5 Merge pull request #9 from stexie99/schemaBranch
590b561 (origin/schemaBranch) initialized post and get routes, changed file names and routes
1d20a95 successfully seeded data to database
eabfae4 updated routes and controllers
3e5f2d6 made changes to file name
f8183ca Merge pull request #8 from stexie99/showBranch
c3ccc81 sending index and show data
932072a Merge pull request #7 from stexie99/movieModel
7a6b368 add movies read route
a595a93 format movie seeder
14f4308 Merge pull request #6 from stexie99/seedermandeep
b13c6c5 Merge branch 'devBranch' into seedermandeep
99eac47 updataseeder
91599da Merge pull request #5 from stexie99/pstarSeeder
7615a30 (origin/pstarSeeder) Merge branch 'devBranch' into pstarSeeder
d617c0b update seeder info
1136182 merging movieSeedData
e75e266 (origin/movieSeedData) make models
e9086bc Merge pull request #3 from stexie99/seederSteven
25febbb (origin/seederSteven) updated seeder data
eb942f8 Merge pull request #2 from stexie99/mikeBranch
e32cfe0 render page
48c3694 (origin/pstarBranch) initiallize devBranch
a6835df testing 2
ed491be testing
defef7f Initial commit


server.js
// modules and globals
require ('dotenv').config()
const express = require('express')
const methodOverride= require('method-override')
const app = express()
const mongoose = require('mongoose')

// middleware
app.set('views', __dirname + '/views')
app.set('view engine', 'jsx')
app.engine('jsx', require('express-react-views').createEngine())
app.use(express.urlencoded({ extended: true }))
app.use(express.json())
app.use(express.static('public'))
app.use(methodOverride('_method'))

// controllers
app.use('/movies', require('./controllers/movie'))

// routes
app.get('/', (req, res) => {
    res.render('home')
})

app.get('*', (req, res) => {
    res.render('Error404')
})

// DB connection
mongoose.connect(process.env.MONGO_URI, { useNewUrlParser: true, useUnifiedTopology: true })
    .then(() => console.log('DB connected'))
    .catch(err => console.error(err));

const PORT = process.env.PORT

app.listen(PORT, console.log(`listening on port ${PORT}`))

controllers/movie.js
const router = require("express").Router()
const Movie = require("../models/movie")
const movieSeedData = require("../models/movieSeedData.js")

//data seeding
router.get("/data/seed", async (req, res) => {
  try {
    await Movie.insertMany(movieSeedData)
    res.redirect("/movies");
  } catch (error) {
    console.error("Error seeding data:", error)
    res.status(500).send("Internal Server Error")
  }
})

//get index page
router.get("/", async (req, res) => {
  const movies = await Movie.find()
  res.render("movies/index", { movies })
})
// create new movie
router.post("/", async (req, res) => {
  try{
    const createMovie = await Movie.create(req.body)
    if(!createMovie){
      res.render("error404")
    }
    res.redirect("movies")
  }catch(err){
    console.log(err)
    res.render("error404")
  }
})
//get new page
router.get("/new", (req, res) => {
  res.render("movies/new")
})
//get edit page
router.get("/:id/edit", async (req, res) => {
  try{
    const { id } = req.params
    const movie = await Movie.findById(id)
    if(!movie){
      res.render("error404")
    }
    res.render("movies/edit", { movie })
  }catch(err){
    console.log(err)
    res.render("error404")
  }
})
//get movie by index
router.get("/:id", async (req, res) => {
  try{
    const movie = await Movie.findById(req.params.id)
    if(!movie){
      res.render("error404")
    }
    res.render("movies/show", { movie })
  }catch(err){
    console.log(err)
    res.render("error404")
  }
})

//edit movie 
router.put("/:id", async (req, res) => {
  const { id } = req.params
  try{
    const movie = await Movie.findByIdAndUpdate(id, req.body)
    if(!movie){
      res.render("error404")
    }
    res.redirect(`/movies/${req.params.id}`)
    }catch(err){
      console.log(err)
      res.render("error404")
    }
})
//delete movie by index
router.delete("/:id", async (req, res) => {
  const { id } = req.params
  try{
    const movie = await Movie.findByIdAndDelete(id)
    if(!movie){
      res.render("error404")
    }
    res.redirect("/movies")
  }catch(err){
    console.log(err)
    res.render("error404")}
})

module.exports = router

models/movieSeedData.js
module.exports = [
  {
    title: "Frozen",
    poster: "https://xl.movieposterdb.com/13_09/2013/2294629/xl_2294629_51b33324.jpg",
    year: "2013",
    director: "Chris Buck",
    genre: "Kids",
  },
  {
    title: "The Lion King",
    poster: "https://xl.movieposterdb.com/22_02/2019/6105098/xl_6105098_3c46c8c0.jpg?v=2023-09-08%2023:39:41",
    year: "1994",
    director: "George Scribner",
    genre: "Musical Drama",
  },
  {
    title: "Zootopia",
    poster: "https://xl.movieposterdb.com/15_11/2016/2948356/xl_2948356_40e450c8.jpg?v=2023-07-26%2004:52:03",
    year: "2016",
    director: "Byron Howard",
    genre: "Action/Comedy",
  },
  {
    title: "Pretty Woman",
    poster: "https://xl.movieposterdb.com/21_02/1990/100405/xl_100405_2c0bd2b6.jpg?v=2023-09-14%2003:00:49",
    year: "1994",
    director: "Garry Marshall",
    genre: "Romance",
  },
  {
    title: "True Lies",
    poster: "https://xl.movieposterdb.com/11_06/1994/111503/xl_111503_ab0d008d.jpg?v=2023-06-16%2021:35:52",
    year: "1994",
    director: "James Cameron",
    genre: "Action",
  },
  {
    title: "The Sixth Sense",
    poster: "https://xl.movieposterdb.com/22_08/1999/167404/xl_167404_b8ed83fd.jpeg",
    year: "1999",
    director: "M. Night Shyamalan",
    genre: "Thriller",
  },
  {
    title: "Spider-Man: Across the Spider-Verse",
    poster: "https://xl.movieposterdb.com/23_06/2022/11290914/xl_spider-man-across-the-spider-verse-part-one-movie-poster_37d41ec0.jpg?v=2023-09-20%2016:45:45",
    year: "2023",
    director: "Joaquim Dos Santos",
    genre: "Action",
  },
  {
    title: "Whiplash",
    poster: "https://xl.movieposterdb.com/20_09/2014/2582802/xl_2582802_5b2aeb39.jpg",
    year: "2014",
    director: "Damien Chazelle",
    genre: "Drama",
  },
  {
    title: "Interstellar",
    poster: "https://xl.movieposterdb.com/15_03/2014/816692/xl_816692_284eb9d5.jpg?v=2023-09-15%2001:37:37",
    year: "2014",
    director: "Christopher Nolan",
    genre: "Science Fiction",
  },
  {
    title: "The Shawshank Redemption",
    poster: "https://xl.movieposterdb.com/05_02/1994/0111161/xl_7266_0111161_d2436dce.jpg?v=2023-01-28%2000:11:44",
    year: "1994",
    director: "Frank Darabont",
    genre: "Drama",
  },
  {
    title: "The Godfather",
    poster: "https://xl.movieposterdb.com/22_07/1972/68646/xl_68646_8c811dec.jpg?v=2023-09-10%2004:02:40",
    year: "1972",
    director: "Francis Ford Coppola",
    genre: "Crime",
  },
  {
    title: "Star Wars: Episode IV - A New Hope",
    poster: "https://xl.movieposterdb.com/09_02/1977/76759/xl_76759_bea34b1e.jpg?v=2023-09-07%2019:43:16",
    year: "1977",
    director: "George Lucas",
    genre: "Action",
  },
  {
    title: "Barbie",
    poster: "https://xl.movieposterdb.com/23_06/2023/1517268/xl_barbie-movie-poster_3cde3986.jpg",
    year: "2023",
    director: "Greta Gerwig",
    genre: "Comedy",
  },
  {
    title: "The Wailing",
    poster: "https://xl.movieposterdb.com/21_11/2016/5215952/xl_5215952_0d5cf387.jpg?v=2022-07-28%2004:44:45",
    year: "2016",
    director: "Na Hong-jin",
    genre: "Horror",
  },
  {
    title: "Gadar 2",
    poster: "https://m.media-amazon.com/images/M/MV5BZGEzMDJjNGUtYTFhZi00MDgyLWIzMzYtMzcwMDQyZjcyNGY1XkEyXkFqcGdeQXVyNTcwNTM5ODI@._V1_FMjpg_UX1000_.jpg",
    year: "2023",
    director: "Anil Sharma",
    genre: "Action",
  },
  {
    title: "The Cannonball Run",
    poster: "https://xl.movieposterdb.com/12_02/1981/82136/xl_82136_e3357ce0.jpg?v=2023-06-04%2007:23:52",
    year: "1981",
    director: "Hal Needham",
    genre: "Action Comedy",
  },
  {
    title: "The Last Unicorn",
    poster: "https://xl.movieposterdb.com/22_08/1982/84237/xl_84237_379e9f1e.jpg?v=2023-03-01%2023:26:50",
    year: "1982",
    director: "Jules Bass",
    genre: "Fantasy",
  },
  {
    title: "Crouching Tiger, Hidden Dragon",
    poster: "https://xl.movieposterdb.com/12_04/2000/190332/xl_190332_75a01a76.jpg?v=2023-02-24%2006:29:23",
    year: "2000",
    director: "Ang Lee",
    genre: "Action",
  },
  {
    title: "Akira",
    poster: "https://xl.movieposterdb.com/05_05/1988/0094625/xl_14935_0094625_99d8108d.jpg?v=2023-07-26%2014:37:08",
    year: "1988",
    director: "Katsuhiro Ôtomo",
    genre: "Animation, Action, Drama",
  },
  {
    title: "Back to the Future",
    poster: "https://xl.movieposterdb.com/13_06/1985/88763/xl_88763_c9357702.jpg",
    year: "1985",
    director: "Robert Zemeckis",
    genre: "Adventure, Comedy, Sci-Fi",
  },
  {
    title: "Bram Stoker's Dracula",
    poster: "https://xl.movieposterdb.com/08_01/1992/103874/xl_103874_40cf7d86.jpg?v=2023-09-20%2020:27:01",
    year: "1992",
    director: "Francis Ford Coppola",
    genre: "Fantasy, Horror",
  },
  {
    title: "The Breakfast Club",
    poster: "https://xl.movieposterdb.com/06_02/1985/0088847/xl_93661_0088847_841b88dc.jpg?v=2023-09-13%2005:20:18",
    year: "1985",
    director: "John Hughes",
    genre: "Comedy, Drama",
  },
  {
    title: "The 'Burbs",
    poster: "https://xl.movieposterdb.com/11_07/1989/96734/xl_96734_3cbfc567.jpg?v=2023-09-20%2018:35:56",
    year: "1989",
    director: "Joe Dante",
    genre: "Comedy, Mystery",
  },
  {
    title: "Donnie Darko",
    poster: "https://xl.movieposterdb.com/11_05/2001/246578/xl_246578_75de6643.jpg?v=2022-07-10%2013:48:01",
    year: "2001",
    director: "Richard Kelly",
    genre: "Drama, Mystery, Sci-Fi",
  },
  {
    title: "The Goonies",
    poster: "https://xl.movieposterdb.com/05_09/1985/0089218/xl_51014_0089218_1e6b1771.jpg?v=2023-09-16%2011:48:49",
    year: "1985",
    director: "Richard Donner",
    genre: "Adventure, Comedy, Family",
  },
  {
    title: "Eternal Sunshine of the Spotless Mind",
    poster: "https://xl.movieposterdb.com/22_12/2004/486419/xl_a-look-inside-eternal-sunshine-of-the-spotless-mind-movie-poster_5a1480ce.jpg?v=2023-09-11%2022:37:11",
    year: "2004",
    director: "Michel Gondry",
    genre: "Drama, Romance, Sci-Fi",
  },
  {
    title: "The Dark Crystal",
    poster: "https://xl.movieposterdb.com/11_04/1982/83791/xl_83791_b4d5ed8b.jpg?v=2023-05-31%2013:01:46",
    year: "1982",
    director: "Jim Henson, Frank Oz",
    genre: "Adventure, Family, Fantasy",
  },
  {
    title: "Paranormal Activity",
    poster: "https://xl.movieposterdb.com/09_11/2007/1179904/xl_1179904_fd57cd33.jpg?v=2023-08-31%2000:31:23",  
    year: "2007",
    director: "Oren Peli",
    genre: "Horror, Mystery",
  },
  {
    title: "Spies Like Us",
    poster: "https://xl.movieposterdb.com/12_05/1985/90056/xl_90056_b83b24ba.jpg?v=2023-05-20%2003:51:31",
    year: "1985",
    director: "John Landis",
    genre: "Adventure, Comedy",
  },
  {
    title: "Doctor Zhivago",
    poster: "https://xl.movieposterdb.com/14_01/1965/59113/xl_59113_cad39c9e.jpg",
    year: "1965",
    director: "David Lean",
    genre: "Drama, Romance, War",
  },
  {
    title: "Kung Fu Panda",
    poster: "https://m.media-amazon.com/images/I/81KE5DERc5L._AC_UF894,1000_QL80_.jpg",
    year: "2008",
    director: "John Stevenson",
    genre: "Kids Comedy Family",
  },
  {
    title: "The Super Mario Bros",
    poster: "https://dx35vtwkllhj9.cloudfront.net/universalstudios/super-mario-bros/images/gallery/image9.jpg",
    year: "2023",
    director: "Aaron Horvath",
    genre: "Adventure",
  },
  {
    title: "Pulp Fiction",
    poster: "https://xl.movieposterdb.com/07_10/1994/110912/xl_110912_55345443.jpg?v=2023-09-22%2016:39:41",
    year: "1994",
    director: "Quentin Tarantino",
    genre: "Crime, Drama",
  },
  {
    title: "Gone with the Wind",
    poster: "https://xl.movieposterdb.com/07_11/1939/31381/xl_31381_035f678b.jpg?v=2023-08-22%2020:25:45",
    year: "1939",
    director: "Victor Fleming",
    genre: "Drama, History, Romance",
  },
  {
    title: "Sully",
    poster: "https://posters.movieposterdb.com/22_04/2016/3263904/l_3263904_430534c9.jpg",
    year: "2016",
    director: "Clint Eastwood",
    genre: "Biography, Drama",
  },
  {
   title: "Out of Africa",
   poster: "https://posters.movieposterdb.com/08_05/1985/89755/s_89755_f14a2bdf.jpg",
   year: "1985",
   director: "Sydney Pollack",
   genre: "Biography, Drama, Romance",
  },
  {
    title: "Lonesome Dove",
    poster: "https://posters.movieposterdb.com/11_09/1989/96639/s_96639_19c1c5ba.jpg",
    year: "1989",
    director: "Simon Wincer",
    genre: "Adventure, Drama, Western",
  },
  {
    title: "Chicken Run",
    poster: "https://posters.movieposterdb.com/12_03/2000/120630/s_120630_247324b5.jpg",
    year: "2000",
    director: "Peter Lord",
    genre: "Animation, Adventure, Comedy",
  },
  {
    title: "Forrest Gump",
    poster: "https://posters.movieposterdb.com/05_06/1994/0109830/s_21293_0109830_af6ba7a1.jpg",
    year: "1994",
    director: "Robert Zemeckis",
    genre: "Drama, Romance",
  },
  {
    title: "Saving Private Ryan",
    poster: "https://posters.movieposterdb.com/07_10/1998/120815/s_120815_e70398d8.jpg",
    year: "1998",
    director: "Steven Spielberg",
    genre: "Drama, War",
  },
  {
    title: "The Devil Wears Prada",
    poster: "https://posters.movieposterdb.com/12_09/2006/458352/s_458352_3b432ee0.jpg",
    year: "2006",
    director: "David Frankel",
    genre: "Comedy, Drama", 
  },
  {
    title: "Charlotte's Web",
    poster: "https://posters.movieposterdb.com/07_10/2006/413895/s_413895_dbdd1585.jpg",
    year: "2006",
    director: "Gary Winick",
    genre: "Comedy, Family, Fantasy",
  },
  {
    title: "Desperado",
    poster: "https://posters.movieposterdb.com/09_01/1995/112851/s_112851_ab6f57a0.jpg",
    year: "1995",
    director: "Robert Rodriguez",
    genre: "Action, Crime, Thriller",
  },
  {
    title: "Armageddon",
    poster: "https://posters.movieposterdb.com/09_06/1998/120591/s_120591_318c008a.jpg",
    year: "1998",
    director: "Michael Bay",
    genre: "Action, Adventure, Sci-Fi",
  },
  {
    title: "The Thomas Crown Affair",
    poster: "https://posters.movieposterdb.com/07_12/1999/155267/s_155267_53cef157.jpg",
    year: "1999",
    director: "John McTiernan",
    genre: "Crime, Romance, Thriller",
  },
  {
    title: "Jurassic Park",
    poster: "https://posters.movieposterdb.com/05_08/1993/0107290/l_45298_0107290_be4e0db3.jpg",
    year: "1993",
    director: "Steven Spielberg",
    genre: "Action, Adventure, Sci-Fi",
  },
  {
    title: "King Kong",
    poster: "https://posters.movieposterdb.com/05_11/2005/0360717/l_68663_0360717_506989dd.jpg",
    year: "2005",
    director: "Peter Jackson",
    genre: "Action, Adventure, Drama",
  }
]
